### PR TITLE
[bidi][js] Allow passing in uri for authentication handlers

### PR DESCRIPTION
**Pull request by pujagani**
### **User description**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Description

Related to #13993

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [X] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Enhanced the `Network` class to support URI-based authentication handlers.
- Introduced a new map `#authHandlers` to store and manage authentication credentials.
- Added a method `getAuthCredentials` to fetch credentials based on the request URI.
- Modified existing methods to integrate the new URI-based authentication logic.
- Added comprehensive tests to verify the functionality of the new authentication handlers with URI filters.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>network.js</strong><dd><code>Add URI-based authentication handlers to Network class</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

javascript/node/selenium-webdriver/lib/network.js

<li>Added support for URI-based authentication handlers.<br> <li> Introduced <code>#authHandlers</code> map to store authentication credentials.<br> <li> Implemented <code>getAuthCredentials</code> method to retrieve credentials based on <br>URI.<br> <li> Modified <code>addAuthenticationHandler</code> and <code>removeAuthenticationHandler</code> <br>methods.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14386/files#diff-47dc7a628e06bb463661104bea1765259446d16694fc05e55e3c89812040e407">+29/-21</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>webdriver_network_test.js</strong><dd><code>Add tests for URI-based authentication handlers</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

javascript/node/selenium-webdriver/test/lib/webdriver_network_test.js

<li>Added tests for authentication handlers with URI filters.<br> <li> Tested multiple handlers with different and same filters.<br> <li> Verified removal and clearing of authentication handlers.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14386/files#diff-d5781b9f5db1cce157f8b2e18e83c6c951e3f7ebf8198244cf57a88880069b96">+42/-1</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: javascript/node/selenium-webdriver/test/lib/webdriver_network_test.js [hunk 1]
```diff
@@ -59,8 +88,20 @@ suite(
         }
       })
 
+      it('can remove authentication handler', async function () {
+        await driver.network().removeAuthenticationHandler(10)
```
Comment (AutomatedTester) [inline javascript/node/selenium-webdriver/test/lib/webdriver_network_test.js]: How do we know it's `10` here ? Looks like we're using magic numbers which isn't ideal for end users

File: javascript/node/selenium-webdriver/lib/network.js [hunk 2]
```diff
@@ -39,39 +40,46 @@ class Network {
...
-  async removeAuthenticationHandler(id) {
+  getAuthCredentials(uri) {
+    for (let [, value] of this.#authHandlers) {
+      if (uri.match(value.uri)) {
+        return value
+      }
+    }
+    return null
+  }
+  async addAuthenticationHandler(username, password, uri = '//') {
```
Comment (AutomatedTester) [inline javascript/node/selenium-webdriver/lib/network.js:67]: We have `uri` as an argument here but when being used later we have things like `basicAuth` and `test`. Are these values pages or types of handlers to expect?

File: javascript/node/selenium-webdriver/test/lib/webdriver_network_test.js (hunk 1 above)
Comment (pujagani) [reply javascript/node/selenium-webdriver/test/lib/webdriver_network_test.js]: I fixed the test and underlying logic. The test is to check if the removing a non existent error throws an error or not.

File: javascript/node/selenium-webdriver/lib/network.js (hunk 2 above)
Comment (pujagani) [reply javascript/node/selenium-webdriver/lib/network.js:67]: "We have uri as an argument here but when being used later we have things like basicAuth and test" - I don't understand this. Here, the URI is the regular expression URI to associate the credentials with.

**Comment by pujagani** [comment]
@AutomatedTester Can we merge this one if everything looks good?

**Review by AutomatedTester — APPROVED** [review]

**Pull request merged**